### PR TITLE
stage2: add more union compile errors / improve error messages

### PR DESCRIPTION
**Pull request by schmee**
To match stage1. For the most part, the error messages are the same, but in some cases I've added some extra information or reformatted it to match how stage2 reports errors.

### CODE REVIEW

**Comment by matu3ba** [comment]
Can you add at least 1 test that shows ~~the difference~~ how awesome this PR is?

**Comment by schmee** [comment]
@matu3ba I'd love to add some compiler error tests for this, unfortunately those tests are not yet enabled for stage2. Once they are, I'll add tests for these things (and more!) 👍

**Comment by andrewrk** [comment]
> @matu3ba I'd love to add some compiler error tests for this, unfortunately those tests are not yet enabled for stage2. Once they are, I'll add tests for these things (and more!) +1

They are - the test cases are in test/compile_errors.zig

**Comment by schmee** [comment]
I'm working on adding compile error tests for this branch and I'm running in to some weird errors in doing so. I'm not sure if it's caused by this PR or if it's exposing some latent bug, investigation is ongoing!

**Comment by schmee** [comment]
So, I got stuck adding the compiler tests. Commit https://github.com/ziglang/zig/pull/11242/commits/ed8b6d873ca37dfb640dbb54a05148f13ad565fb has 5 different tests which all pass individually, but when ran together produces [this error](https://gist.github.com/schmee/7cbc2c54be51fd826ab321373a321b18) (on macOS x86_64). I've tried various permutations of different tests together, some work and other don't, and I can't figure out the pattern. I also tried to reproduce the error on master without success.

In short, I'm not able to move this PR forward at the moment if adding these tests is a requirement for merging. I'll work on something else for a while and return to this later when I have some new ideas on how to tackle it.

@andrewrk this is the PR I mentioned on Discord, in case you have time to look at it at some point 🙏

**Comment by schmee** [comment]
Waiting for https://github.com/ziglang/zig/pull/11284 to get merged before I return to this 👍

**Comment by schmee** [comment]
@andrewrk seems like whatever was causing the tests to fail has been resolved in master, so this is now ready for review 👍

File: src/Sema.zig [hunk 1]
```diff
@@ -22055,6 +22103,10 @@ fn semaUnionFields(mod: *Module, union_obj: *Module.Union) CompileError!void {
         } else {
             // The provided type is the enum tag type.
             union_obj.tag_ty = try provided_ty.copy(decl_arena_allocator);
+            // The fields of the union must match the enum exactly.
+            // Store a copy of the enum field names so we can check for
+            // missing or extraneous fields later.
+            tag_ty_field_names = try union_obj.tag_ty.enumFields().clone(decl_arena_allocator);
```
Comment (Vexu) [inline src/Sema.zig:22109]: The cloned map doesn't seem to leave the function so this should use `sema.arena` or `gpa`.

**Review by Vexu — APPROVED** [review]

File: src/Sema.zig [hunk 2]
```diff
@@ -22148,6 +22200,20 @@ fn semaUnionFields(mod: *Module, union_obj: *Module.Union) CompileError!void {
             set.putAssumeCapacity(field_name, {});
         }
 
+        if (tag_ty_field_names) |*names| {
+            const enum_has_field = names.contains(field_name);
```
[the remark this answers is not in the document]
Comment (Vexu) [reply src/Sema.zig:22204]: Also regardless of the above suggestion, shouldn't this come after the duplicate field check?
```zig
const E = enum {a, b};
const U = union(E) {
    a: u32,
    a: u32, // should be duplicate field error, enum 'E' has no field named 'a'
};
```

[the remark this answers is not in the document]
Comment (schmee) [reply src/Sema.zig]: You're absolutely right! These two checks were originally two different PRs but when I merged them they ended up in the wrong order, sorry bout that. Fixed!

**Comment by schmee** [comment]
Thanks for the review @Vexu! I have addressed your comments.
